onvm/onvm_nflib/onvm_common.h

```diff
@@ -211,71 +211,92 @@ struct core_status {
 /* Function prototype for NFs that want extra initalization/setup before running */
-typedef void (*setup_func)(struct onvm_nf_context *nf_context);
+typedef void (*setup_func)(struct onvm_nf_local_ctx *nf_local_ctx);
 /* Function prototype for NFs to handle custom messages */
-typedef void (*handle_msg_func)(void *msg_data, struct onvm_nf_info *nf_info);
+typedef void (*handle_msg_func)(void *msg_data, struct onvm_nf *nf);
 /* Function prototype for NFs to signal handling */
 typedef void (*handle_signal_func)(int);
 
 /* Information needed to initialize a new NF child thread */
 struct onvm_nf_scale_info {
-        struct onvm_nf_info *parent;
-        uint16_t instance_id;
-        uint16_t service_id;
-        uint16_t core;
-        uint8_t flags;
-        char *tag;
-        void *data;
-        setup_func setup_func;
-        pkt_handler_func pkt_func;
-        callback_handler_func callback_func;
-        advanced_rings_func adv_rings_func;
-        handle_msg_func handle_msg_function;
+        struct onvm_nf_init_cfg *nf_init_cfg;
+        struct onvm_nf *parent;
+        void * data;
+        struct {
+                setup_func setup;
+                handle_msg_func handle_msg;
+                callback_func callback;
+                pkt_handler_func pkt_handler;
+                /* Deprecated, will be removed in the future advanced rings rework */
+                advanced_rings_func adv_rings;
+        } functions;
 };
 
-struct onvm_nf_context {
+struct onvm_nf_local_ctx {
         struct onvm_nf *nf;
-        struct onvm_nf_info *nf_info;
         rte_atomic16_t nf_init_finished;
         rte_atomic16_t keep_running;
 };
 
 /*
- * Define a nf structure with all needed info, including
- * stats from the nfs.
+ * Define a NF structure with all needed info, including:
+ *      thread information, function callbacks, flags, stats and shared core info.
+ *
+ * This structure is available in the NF when processing packets or executing the callback.
  */
 struct onvm_nf {
         struct rte_ring *rx_q;
         struct rte_ring *tx_q;
         struct rte_ring *msg_q;
-        struct onvm_nf_info *info;
-        uint16_t instance_id;
-        /* Advanced ring mode or packet handler mode */
-        uint8_t nf_mode;
-        /* Instance ID of parent NF or 0 */
-        uint16_t parent;
-        rte_atomic16_t children_cnt;
         /* Struct for NF to NF communication (NF tx) */
         struct queue_mgr *nf_tx_mgr;
-        /* Pointer to NF context (used for signal handling/termination */
-        struct onvm_nf_context *context;
+        uint16_t instance_id;
+        uint16_t service_id;
+        uint8_t status;
+        /* Deprecated, will be removed in the future advanced rings rework */
+        uint8_t nf_mode;
+        char *tag;
+        /* Pointer to NF defined state data */
+        void *data;
+        /* Pointer to NF context (used for signal handling/terminating NF's children) */
+        struct onvm_nf_local_ctx *context;
```

Comment:
Why does the `onvm_nf` struct have a reference to `onvm_nf_local_ctx` and vice versa?  As in, is there a circular reference between both of these structs for an NF?

Reply:
@nks5295 When we're doing children cleanup we need to basically set all keep_running variables in nf_context to 0. This allows us to achieve that